Modules/mathmodule.c

```diff
@@ -1476,6 +1476,194 @@ count_set_bits(unsigned long n)
+        """
+        n = operator.index(n)
+
+        if n < 0:
+            raise ValueError("isqrt() argument must be nonnegative")
+        elif n == 0:
+            return 0
+
+        c = (n.bit_length() - 1) // 2
+        d = 0
+        a = 1
+        for s in reversed(range(c.bit_length())):
+            e = d
+            d = c >> s
+            a = (a << d - e - 1) + (n >> 2*c - e - d + 1) // a
+            # Loop invariant
+            assert (a-1)**2 < n >> 2*(c - d) < (a+1)**2
+
+        return a - (a*a > n)
+
+*/
+
+/*[clinic input]
+math.isqrt
+
+    n : 'O'
+    /
+
+Return the largest integer not exceeding the square root of the input.
+[clinic start generated code]*/
+
+static PyObject *
+math_isqrt(PyObject *module, PyObject *n)
+/*[clinic end generated code: output=35a6f7f980beab26 input=633fa99eb56c4045]*/
+{
+    int a_too_large, s;
+    size_t n_bit_length, c, d, e;
+    PyObject *a=NULL, *b, *q=NULL, *shift;
+
+    n = PyNumber_Index(n);
+    if (n == NULL) {
+        return NULL;
+    }
+
+    if (_PyLong_Sign(n) < 0) {
+        PyErr_SetString(
+            PyExc_ValueError,
+            "isqrt() argument must be nonnegative");
+        goto error;
+    }
+    if (_PyLong_Sign(n) == 0) {
+        Py_DECREF(n);
+        a = _PyLong_Zero;
+        Py_INCREF(a);
+        return a;
+    }
+
+    n_bit_length = _PyLong_NumBits(n);
+    if (n_bit_length == (size_t)(-1)) {
+        goto error;
+    }
+
+    c = (n_bit_length - 1U) / 2U;
+    s = 0;
+    while ((c >> s) > 0) {
+        ++s;
+    }
+
+    a = _PyLong_One;
+    Py_INCREF(a);
+
+    d = 0;
+    while (s > 0) {
+        --s;
+
+        e = d;
+        d = c >> s;
+
+        /* q = (n >> 2*c - e - d + 1) // a */
+        shift = PyLong_FromSize_t(2U*c - d - e + 1U);
+        if (shift == NULL) {
+            goto error;
+        }
+        b = PyNumber_Rshift(n, shift);
+        Py_DECREF(shift);
+        if (b == NULL) {
+            goto error;
+        }
+        q = PyNumber_FloorDivide(b, a);
+        Py_DECREF(b);
+        if (q == NULL) {
+            goto error;
+        }
+
+        /* a = (a << d - 1 - e) + q */
+        shift = PyLong_FromSize_t(d - 1U - e);
+        if (shift == NULL) {
+            goto error;
+        }
+        b = PyNumber_Lshift(a, shift);
+        Py_DECREF(a);
+        Py_DECREF(shift);
+        if (b == NULL) {
+            goto error;
+        }
+        a = PyNumber_Add(b, q);
+        Py_DECREF(b);
+        Py_DECREF(q);
+        if (a == NULL) {
+            goto error;
+        }
+    }
+
+    /* The correct result is either a or a - 1. Figure out which, and
+       decrement a if necessary. */
+
+    /* a_too_large = n < a * a */
+    b = PyNumber_Multiply(a, a);
+    if (b == NULL) {
+        goto error;
+    }
+    a_too_large = PyObject_RichCompareBool(n, b, Py_LT);
+    Py_DECREF(b);
+    if (a_too_large == -1) {
+        goto error;
+    }
+
+    if (a_too_large) {
+        b = PyNumber_Subtract(a, _PyLong_One);
```

Comment:
In one line:
`Py_SETREF(a, PyNumber_Subtract(a, _PyLong_One));`

Reply:
Ah, that's definitely nicer. Will change.